circus-train-gcp/src/test/java/com/hotels/bdp/circustrain/gcp/context/GCPBeanPostProcessorTest.java

```diff
@@ -16,46 +16,78 @@
 package com.hotels.bdp.circustrain.gcp.context;
 
 import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.doReturn;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 
 import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
-import org.mockito.Spy;
+import org.mockito.Mock;
 import org.mockito.runners.MockitoJUnitRunner;
 
 import com.hotels.bdp.circustrain.context.CommonBeans;
+import com.hotels.bdp.circustrain.gcp.BindGoogleHadoopFileSystem;
+import com.hotels.bdp.circustrain.gcp.DistributedFileSystemPathProvider;
+import com.hotels.bdp.circustrain.gcp.FileSystemFactory;
+import com.hotels.bdp.circustrain.gcp.GCPCredentialCopier;
+import com.hotels.bdp.circustrain.gcp.GCPCredentialPathProvider;
+import com.hotels.bdp.circustrain.gcp.RandomStringFactory;
 
 @RunWith(MockitoJUnitRunner.class)
 public class GCPBeanPostProcessorTest {
-  private @Spy GCPSecurity security = new GCPSecurity();
-  private @Spy Configuration configuration = new Configuration();
+  private @Mock RandomStringFactory randomStringFactory;
+
+  private @Mock GCPSecurity security;
+  private @Mock GCPCredentialPathProvider credentialPathProvider;
+  private @Mock DistributedFileSystemPathProvider distributedFileSystemPathProvider;
+  private @Mock Configuration configuration;
+  private @Mock BindGoogleHadoopFileSystem bindGoogleHadoopFileSystem;
+  private @Mock FileSystemFactory fileSystemFactory;
+  private @Mock GCPCredentialCopier credentialCopier;
+
+  private GCPBeanPostProcessor processor;
+
+  @Before
+  public void init() {
+    doNothing().when(configuration).set(anyString(), anyString());
+    processor = new GCPBeanPostProcessor(credentialPathProvider, distributedFileSystemPathProvider,
+        bindGoogleHadoopFileSystem, fileSystemFactory, credentialCopier);
+  }
 
   @Test
   public void postProcessAfterInitializationWithCorrectBeanName() throws Exception {
```

Comment:
I am not a fan of this test because it is not really testing what the code is supposed to do, you're only testing that a `newPath` is called but that's not the whole picture.
Just going on the branches in the code , you should have three tests:
1) a test where the bean name doesn't match, you got that one (`postProcessAfterInitializationWithIncorrectBeanName`), good!
2) a test where bean is good and path is not null (almost got that `postProcessorWithNonNullCredentialProvider`)
3) a test where bean is good and path is null (almost got that `postProcessAfterInitializationWithCorrectBeanName`)

I think they can be a bit more clear and verify better what the code is doing.
Consider what you have now for 2,3 and replace with:
```
  @Test
  public void postProcessAfterInitializationWithConfigurationBeanProviderPathIsNotNull() throws Exception {
    String beanName = CommonBeans.BEAN_BASE_CONF;
    when(credentialPathProvider.newPath()).thenReturn(new Path("/test.json"));
    when(fileSystemFactory.getFileSystem(configuration)).thenReturn(fileSystem);

    processor.postProcessAfterInitialization(configuration, beanName);
    verify(bindGoogleHadoopFileSystem).bindFileSystem(configuration);
    verify(credentialCopier)
        .copyCredentials(fileSystem, configuration, credentialPathProvider, distributedFileSystemPathProvider);
  }

  @Test
  public void postProcessAfterInitializationWithConfigurationBeanProviderPathIsNull() throws Exception {
    String beanName = CommonBeans.BEAN_BASE_CONF;
    when(credentialPathProvider.newPath()).thenReturn(null);

    processor.postProcessAfterInitialization(configuration, beanName);
    verifyZeroInteractions(bindGoogleHadoopFileSystem, credentialCopier, fileSystemFactory);
  }
```


If you prefer `doReturn.when` instead of `when.thenReturn` I am ok with that, you'll see the latter a bit more in our code though (as you get compile time type checking and imho it reads better.) I am sure there is a whole discussion somewhere on the internet favoring one or the other. Either way I can read it.